README.md
# .Net Image Library

.NET library for easy image handling like:
  * Filters
  * Cropping
  * Thumbnail creation
  * Saving as JPG/GIF/PNG files or streams
  * Opening images from streams or URLs

You'll find information on how to get started at http://kaliko.com/image-library/get-started/ and full API documentation at http://kaliko.com/image-library/api/

<a href="https://www.nuget.org/packages/ImageLibrary/">Get .Net Image Library from NuGet</a>

Using only safe code making this library possible to use on web hosts with medium trust.

When having access to full trust make sure to use ImageLibrary.FastFilters to greatly improve performance!

Current build contains the following filters:
  * Gaussian blur filter
  * Unsharpen mask filter
  * Chroma key filter
  * Contrast filter
  * Brightness filter
  * Invert filter
  * Desaturnation filter

If you plan using this library with WPF or simular, read this post on <a href="http://labs.kaliko.com/2011/03/convert-to-bitmapimage.html">how to convert an KalikoImage object to System.Windows.Media.Imaging.BitmapImage and System.Windows.Controls.Image</a>.

## Examples

### Crop scaling
Scale by cropping the image to cover the requested width and height:
```csharp
    using (var image = new KalikoImage(ImageFilePath)) {
        image
            .Scale(new CropScaling(500, 500))
            .SaveJpg(OutputFilePath, 80);
    }
```

### Fit scaling
Scale by fitting the image inside the requested width and height:
```csharp
    using (var image = new KalikoImage(ImageFilePath)) {
        image
            .Scale(new FitScaling(500, 500))
            .SaveJpg(OutputFilePath, 80);
    }
```

### Pad scaling
Scale by fitting the image inside the requested width and height and then pad to the width and height using colour:
```csharp
    using (var image = new KalikoImage(ImageFilePath)) {
        image
            .Scale(new PadScaling(500, 500, Color.Crimson))
            .SaveJpg(OutputFilePath, 80);
    }
```

### Focal point scaling
Similar to the crop scaling but takes a focal point as parameter to determine the point of interest when cropping the image:
```csharp
    using (var image = new KalikoImage(ImageFilePath)) {
        image
            .Scale(new FocalPointScaling(500, 500, 0.75, 1))
            .SaveJpg(OutputFilePath, 80);
    }
```

### Brightness filter
```csharp
    using (var image = new KalikoImage(ImagePath)) {
        image.ApplyFilter(new FastBrightnessFilter(50));
        image.SaveJpg(OutputPath, 80);
    }
```

### Contrast filter
```csharp
    using (var image = new KalikoImage(ImagePath)) {
        image.ApplyFilter(new FastContrastFilter(50));
        image.SaveJpg(OutputPath, 80);
    }
```

### Chroma key filter (green screen)
```csharp
    using (var image = new KalikoImage("c:\\images\\greenscreen.jpg"))) {
        // Make green screen parts transparent
        image.ApplyFilter(new FastChromaKeyFilter(Color.FromArgb(13, 161, 37), 40, 0.5f, 0.75f));

        using (var background = new KalikoImage("c:\\images\\backdrop.jpg")) {
            // Apply cleaned green screen image over backdrop
            background.BlitImage(image);
            background.SaveJpg(OutputPath, 80);
        }
    }
```

### Desaturate filter
```csharp
    using (var image = new KalikoImage(ImagePath)) {
        image.ApplyFilter(new FastDesaturationFilter());
        image.SaveJpg(OutputPath, 80);
    }
```

### Gaussian blur filter
```csharp
    using (var image = new KalikoImage(ImagePath)) {
        image.ApplyFilter(new FastGaussianBlurFilter(1.1f));
        image.SaveJpg(OutputPath, 80);
    }
```

### Invert filter
```csharp
    using (var image = new KalikoImage(ImagePath)) {
        image.ApplyFilter(new FastInvertFilter());
        image.SaveJpg(OutputPath, 80);
    }
```

### Unsharp mask filter
```csharp
    using (var image = new KalikoImage(ImagePath)) {
        image.ApplyFilter(new FastUnsharpMaskFilter(1.1f, 0.44f, 0));
        image.SaveJpg(OutputPath, 80);
    }
```

### Draw text
```csharp
    using (var image = new KalikoImage(ImagePath)) {
        var text = new TextField("Lorem ipsum") {
            Alignment = StringAlignment.Center,
            VerticalAlignment = StringAlignment.Center,
            Outline = 5,
            OutlineColor = Color.Red,
            Font = new Font("Arial", 60),
            Rotation = 30f,
            TextColor = Color.DarkOrange,
            TextShadow = new TextShadow(Color.FromArgb(128, 0, 0, 0), 4, 4)
        };
        image.DrawText(text);
        image.SaveJpg(OutputPath, 80);
    }
```

### Gradient fill
```csharp
    using (var image = new KalikoImage(ImagePath))
    {
        image.GradientFill(Color.FromArgb(128, 255, 200, 90), Color.FromArgb(128, 255, 64, 0));
        image.SaveJpg(OutputPath, 80);
    }
```

### Watermark
```csharp
    using (var image = new KalikoImage(ImagePath))
    {
        image.BlitFill("c:\\images\\semi-transparent-watermark.png"));
        image.SaveJpg(OutputPath, 80);
    }
```


## History

**4.0.0**
* Migrated projects to be compatible with .NET 4.x and Core
* Added FocalPointScaling as an alternative to crop scaling but that allows to define the point of interest

**3.0.2**
* Fixed bug where BlitFill doesn't properly fill portrait sized images

**3.0.1**
* Added option to prevent upscaling when resizing images

**3.0.0**
* Added faster filter alternatives for full trust environmnets
* Added SetResolution functions

**2.0.6**
* Added support to store original or set new resolution (DPI) 
* Added RotateFlip method
* Added access to bitmap data

**2.0.5**
* Rewritten file loader to prevent file locks
* Fixed image loading to ignore pixel-per-inch resolutions of original images
* Fixed constructor using System.Drawing.Image to support indexed palettes

**2.0.4**
* Added new TextField class for better text support
* Fixed scaling bug and updated test program 

**2.0.0**
  * Replaced Gaussian blur filter with better implementation (affects unsharpen masks)
  * Added chroma key filter
  * Rewritten API for Scaling
  * Added color space handling

**1.2.4**
  * Updated to Visual Studio 2010.
  * Code clean-up.<br>
  * Unwanted-border-artifact-problem fixed (thanks Richard!)<br>
  * IDisponable has been implemented.<br>

**1.2.3**
  * Minor changes.
  * First API documentation uploaded. Still missing a whole lot, but it's a start :)

**1.2.2**
  * Minor changes

**1.2.1**
  * Bug in thumbnail function fixed.
  * Code cleaned up.


ReleaseNotes.md
### New in 4.0.0
* Migrated projects to be compatible with .NET 4.x, Core 2.x and Standard 2.x
* Added FocalPointScaling as an alternative to crop scaling but that allows to define the point of interest
* Thanks to rbrian for pull request!

### New in 3.0.2
* Fixed bug where BlitFill doesn't properly fill portrait sized images

### New in 3.0.1
* Added option to prevent upscaling when resizing images

### New in 3.0.0
* Added faster filter alternatives for full trust environmnets
* Added SetResolution functions

### New in 2.0.6
* Added support to store original or set new resolution (DPI) 
* Added RotateFlip method
* Added access to bitmap data

### New in 2.0.5
* Rewritten file loader to prevent file locks
* Fixed image loading to ignore pixel-per-inch resolutions of original images
* Fixed constructor using System.Drawing.Image to support indexed palettes

### New in 2.0.4
* Added new TextField class for better text support
* Fixed scaling bug and updated test program 

### New in 2.0.0
* Replaced Gaussian blur filter with better implementation (affects unsharpen masks)
* Added chroma key filter
* Rewritten API for Scaling
* Added color space handling

### New in 1.2.4
* Updated to Visual Studio 2010.
* Code clean-up.
* Unwanted-border-artifact-problem fixed (thanks Richard!)
* IDisponable has been implemented.

### New in 1.2.3
* Minor changes.
* First API documentation uploaded. Still missing a whole lot, but it's a start :)

### New in 1.2.2
* Minor changes

### New in 1.2.1
* Bug in thumbnail function fixed.
* Code cleaned up.


TestApp/Program.cs
using System;
using System.Drawing;
using System.Drawing.Imaging;
using Kaliko.ImageLibrary;
using Kaliko.ImageLibrary.Filters;
using System.IO;
//using Kaliko.ImageLibrary.Transformation;

namespace TestApp {
    using System.Threading;

    class Program {
        private static string _failed = string.Empty;

        static void Main() {
            var dateTime = DateTime.Now;

/*            var image = new KalikoImage(@"testimage.jpg");
            image.SetFont("84_rock.ttf", 120, FontStyle.Regular);
            image.Color = Color.FromArgb(64, Color.White);
            image.WriteText("Hello World!", 0, 0);
            image.SaveJpg("Output.jpg", 90);*/


            //var image = new KalikoImage(@"Premiere-Pro-chroma-key-original.jpg");
            //image.ApplyFilter(new ChromaKeyFilter(Color.FromArgb(120,209,129), 35f, 0.1f, 0.2f));
            //image.SavePng("Output-ChromaKeyed.png", 90);


            var image = new KalikoImage(@"Premiere-Pro-chroma-key-original.jpg");
            //image.ApplyFilter(new GaussianBlurFilter(10));
            //image.SavePng("Output-gaussian-10.png", 90);
            //image = new KalikoImage(@"Premiere-Pro-chroma-key-original.jpg");
            //image.ApplyFilter(new GaussianBlurFilter(20));
            //image.SavePng("Output-gaussian-20.png", 90);
            //image = new KalikoImage(@"Premiere-Pro-chroma-key-original.jpg");
            //image.ApplyFilter(new GaussianBlurFilter(30));
            //image.SavePng("Output-gaussian-30.png", 90);
            //image = new KalikoImage(@"Premiere-Pro-chroma-key-original.jpg");
            //image.ApplyFilter(new GaussianBlurFilter(40));
            //image.SavePng("Output-gaussian-40.png", 90);

            image = new KalikoImage(@"Premiere-Pro-chroma-key-original.jpg");
            image.ApplyFilter(new UnsharpMaskFilter(2.2f, 0.4f, 0));
            image.SavePng("Output-gaussian-unsharp.png");


            image = new KalikoImage(@"Premiere-Pro-chroma-key-original.jpg");
            image.ApplyFilter(new GaussianBlurFilter(2.2f));
            image.SavePng("Output-gaussian.png");

            //image.ApplyFilter(new BlueFilter());
            //image.ApplyFilter(new NewGaussian(30));
            //image.ApplyFilter(new GaussianBlurFilter(30));
            //image.SaveJpg("Output.jpg", 90);


            Console.WriteLine("Time taken " + (DateTime.Now-dateTime).TotalMilliseconds +" ms");
            Thread.Sleep(2000);
/*
    // Open image from file system
    var image = new KalikoImage(@"C:\MyImages\Image.png");
            
    // Open image from a stream
    MemoryStream memoryStream = ...
    var image = new KalikoImage(memoryStream);
            
    // Creating a new empty image
    var image = new KalikoImage(640, 480, Color.White);

    // Save image to file system in jpg format with quality setting 90
    image.SaveJpg(@"C:\MyImages\Output.jpg", 90);
            
    // Save image to stream in jpg format with quality setting 90
    MemoryStream memoryStream = new MemoryStream();
    image.SaveJpg(memoryStream, 90);

    // Send the image in jpg format as a HttpResponse
    image.StreamJpg(80, "MyImage.jpg");

    // Save image to file system in the selected format
    image.SaveImage(@"C:\MyImages\Output.tif", ImageFormat.Tiff);
*/
/*            image = new KalikoImage(512, 512);

    // Load the font and set both font size as well as type
    image.SetFont("84_rock.ttf", 80, FontStyle.Regular);
    
    // Set the color that will be used, a semi-transparent white in this case        
    image.Color = Color.FromArgb(60, Color.White);

    // Write the text
    image.WriteText("My string", 0, 0);
            

    // Place the source image on top, left of our image
    var sourceImage = new KalikoImage(@"C:\Img\Stamp.png");
    image.BlitImage(sourceImage, 0, 0);

    // Repeat the above, but in a single call
    image.BlitImage(@"C:\Img\Stamp.png", 0, 0);


    // Create a new image and fill the source image all over
    var image = new KalikoImage(640, 480);
    var patternImage = new KalikoImage(@"C:\Img\Checkered.png");
    image.BlitFill(patternImage);

    // Repeat the above, but in a just one additional call
    var image = new KalikoImage(640, 480);
    image.BlitFill(@"C:\Img\Checkered.png");
            
            image.SaveJpg("test.jpg", 99);


            KalikoImage sharpimg = image.GetThumbnailImage(100, 100, ThumbnailMethod.Crop);
            sharpimg.ApplyFilter(new UnsharpMaskFilter(1.2, 0.3));
            sharpimg.SaveJpg("thumbnail-unsharpened.jpg", 90);

            sharpimg.ApplyFilter(new DesaturationFilter());
            sharpimg.SaveJpg("thumbnail-gray.jpg", 90);

            sharpimg.ApplyFilter(new ContrastFilter(30));
            sharpimg.SaveJpg("thumbnail-contrast.jpg", 90);
            //PixelFormatTest();
            */
            /*

            var fileStream = new FileStream("testimage.jpg", FileMode.Open, FileAccess.Read);
            byte[] imgBytes = new byte[fileStream.Length];
            fileStream.Read(imgBytes, 0, (int)fileStream.Length);
            fileStream.Close();

            KalikoImage img = new KalikoImage(new MemoryStream(imgBytes));

            img.BlitFill(@"alpha.png");

            MemoryStream ms = new MemoryStream();
            img.Image.Save(ms, System.Drawing.Imaging.ImageFormat.Jpeg);
            byte[] array = ms.ToArray();

            var stream = new FileStream("testXX.jpg", FileMode.Create);
            stream.Write(array, 0, array.Length);
            stream.Close();

            */

            //KalikoImage dest = new KalikoImage(640, 480);
            /*
            Matrix matRot = new Matrix();
            Matrix matTrans1 = new Matrix();
            Matrix matTrans0 = new Matrix();
            Matrix matScale = new Matrix();
            matTrans0.Translation(-(float)(image.Width) / 2.0f, -(float)(image.Height) / 2.0f);
            matRot.Rotation(0f);
            matScale.Scaling(0.5f, 0.5f);
            matTrans1.Translation((float)(dest.Width) / 2.0f, (float)(dest.Height) / 2.0f);


            Transforms.ImageTransform(dest, image, matTrans1 * ( matScale * (matRot * matTrans0)), Transforms.NearestNeigbour);

            dest.SaveJpg("s:\\aaa.jpg", 99);
            */

/*            image.SetFont("84_rock.ttf", 40, FontStyle.Regular);
            image.Color = Color.White;
            image.WriteText("A TEST OF STATIC", 200, 100);
            image.WriteText("A TEST OF ROTATION", 200, 100, 45);
            image.Color = Color.OrangeRed;
            image.WriteText("A TEST OF ROTATION", 200, 100, 15);
            image.WriteText("A TEST OF ROTATION", 200, 100, 80);
            */
            //image.SaveJpg("test.jpg", 100);

            //image.GetThumbnailImage(430, 430, ThumbnailMethod.Fit).SaveJpg("thumbnail-fitx.jpg", 90);

            //image.GetThumbnailImage(100, 140, ThumbnailMethod.Crop).SaveJpg("thumbnail-crop.jpg", 90);

            //image.GetThumbnailImage(100, 140, ThumbnailMethod.Pad).SaveJpg("thumbnail-pad.jpg", 90);

            /*
            KalikoImage sharpimg = image.GetThumbnailImage(100, 100, ThumbnailMethod.Crop);
            sharpimg.ApplyFilter(new UnsharpMaskFilter(1.2, 0.3));
            sharpimg.SaveJpg("thumbnail-unsharpened.jpg", 90);

            sharpimg.ApplyFilter(new DesaturationFilter());
            sharpimg.SaveJpg("thumbnail-gray.jpg", 90);

            sharpimg.ApplyFilter(new ContrastFilter(30));
            sharpimg.SaveJpg("thumbnail-contrast.jpg", 90);
            */



        }

        private static void PixelFormatTest() {
            TestFormat(PixelFormat.Format16bppGrayScale);
            TestFormat(PixelFormat.Format16bppRgb555);
            TestFormat(PixelFormat.Format16bppArgb1555);
            TestFormat(PixelFormat.Format16bppRgb565);
            TestFormat(PixelFormat.Format1bppIndexed);
            TestFormat(PixelFormat.Format24bppRgb);
            TestFormat(PixelFormat.Format32bppArgb);
            TestFormat(PixelFormat.Format32bppPArgb);
            TestFormat(PixelFormat.Format32bppRgb);
            TestFormat(PixelFormat.Format48bppRgb);
            TestFormat(PixelFormat.Format4bppIndexed);
            TestFormat(PixelFormat.Format64bppArgb);
            TestFormat(PixelFormat.Format64bppPArgb);
            TestFormat(PixelFormat.Format8bppIndexed);

            string a = _failed;

            /*
             * OutOfMemoryException:Format16bppGrayScale
             * OutOfMemoryException:Format16bppArgb1555
             * Exception: Format1bppIndexed
             * Exception: Format4bppIndexed
             * Exception: Format8bppIndexed
             */
        }

        private static void TestFormat(PixelFormat pixelFormat) {
            Bitmap test = new Bitmap(200, 200, pixelFormat);
            try {
                Graphics gr = Graphics.FromImage(test);
            }
            catch (OutOfMemoryException e) {
                _failed += "OutOfMemoryException:" + pixelFormat + "\r\n";
            }
            catch (Exception e) {
                _failed += "Exception: " + pixelFormat + "\r\n";
            }

        }
    }
}
